### PR TITLE
Improve the derivative of the QR decomposition

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76115

We derive and implement a more concise rule for the forward and backward
derivatives of the QR decomposition. While doing this we:
- Fix the composite compliance of `linalg.qr` and we make it support functorch's batching transform
- Add preconditions to the forward AD.
- Correct / improve the docs, as they contained some inaccuracies.
- Improve the performance and simplify the implementation of both forward and backward
- Avoid saving the input matrix for the backward computation.
- Activate the tests when there's no magma but there's cusolver.

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -3155,170 +3155,171 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
...
-    } else {
-      // when grad_R is not defined then grad_V and grad_Q_prime
-      // get initialized with zeros
-      grad_V = at::zeros_like(Y, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-      grad_Q_prime = at::zeros_like(q, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+      gA = gR.matmul(R.mH());
+  }
+  if (m >= n) {
+    const auto syminvadj = [](const Tensor& X) {
+      return X + X.mH() - at::diag_embed(at::real(X.diagonal(0, -2, -1)));
```
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:3297]: @soulitzer  If you change this by, say
```c++
auto ret = X + X.mH();
at::real(ret.diagonal(0,-2, -1)).mul_(0.5);
return ret;
```
the composite compliance test breaks.

Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:3297]: Ok the issue seems to be that when we wrap the output in CompositeCompliantTensor, `_make_wrapper_subclass` ignores the strides if it sees that the size is zero and always sets it to 1.

```
elem = torch.ones(0,).as_strided((0,), (2,), 0)
ret = CompositeCompliantTensor(elem)
print(f"{ret.shape = }, {Tensor.stride(ret) = }, {ret.elem.shape = }, {ret.elem.stride() = }")
```

Prints:
```
ret.shape = torch.Size([0]), Tensor.stride(ret) = (1,), ret.elem.shape = torch.Size([0]), ret.elem.stride() = (2,)
```

Either we fix `make_wrapper_subclass` or update CompositeCompliantTensor to skip the check for strides when size is 0.

cc @zou3519

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -3155,170 +3155,171 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
...
+  // dQ = dAR^{-1} - QdRR^{-1}
+  // Then we have
+  // Q^H dA R^{-1} = Q^HdQ + dRR^{-1}
+  // where Q^HdQ is skew Hermitian and dRR^{-1} is upper triangular
+  // Define sym(X) = X + X^H
+  // sym(dRR^{-1}) = sym(Q^H dA R^{-1})
+  // and define syminv(X) = triu(X) - 0.5 * diag(X) the inverse of sym : Triu(k) -> Her(k) to give
+  // dR = syminv(sym(Q^H dA R^{-1}))R
+  //
+  // Case m <= n
```
Comment (vadimkantorov) [inline torch/csrc/autograd/FunctionsManual.cpp:3174]: `m < n`?

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:3174]: You mean `m > n`?

Comment (vadimkantorov) [reply torch/csrc/autograd/FunctionsManual.cpp:3174]: The other case checks `m >= n` in the code and in the comment, the other case is `else`. So I'm just proposing consistency between the code and the comment and clearness which branch is followed for `m == n`

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
...
+  }
 }
 
-Tensor linalg_qr_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
-                          c10::string_view mode, const Tensor& q, const Tensor& r){
+Tensor linalg_qr_backward(const Tensor& gQ, const Tensor& gR,
+                          const Tensor& Q, const Tensor& R,
+                          const c10::string_view mode) {
+  // Nb. We won't be too formal below, as writing this proof formaly is a pain
+  // We'll link here a formal writing of all this at some point in the future
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3245]: I wrote down the proof of the formulas, in line with most of the other implementations of forward/backward linalg formulas in PyTorch.

The point of this note is that the formulas derived in those links are not formal (even more, the first one refers to another preprint for the derivation). Even more, they just provide the formulas for the case m >= n. I wrote down the derivation for the case m <= n (m < n in the implementation) as well, so this should do for now. The second line of the comment points out that, once the document I'm writing is ready, I'll replace all the derivations with links to this comprehensive paper, which should present all these formulas them in a formal and concise way.

[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3245]: Offline sync. It's good to replace the references by the actual derivation to have the code and the comment explaining the code in sync.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3303]: fwiw, once this problem is fixed, I think that a number of `linalg` functions will start being composite compliant. For exmaple the eigenvalue decompositions.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -11238,13 +11238,13 @@
   python_module: linalg
   variants: function
 
-- func: linalg_qr(Tensor self, str mode='reduced') -> (Tensor Q, Tensor R)
+- func: linalg_qr(Tensor A, str mode='reduced') -> (Tensor Q, Tensor R)
```
Comment (nikitaved) [inline aten/src/ATen/native/native_functions.yaml:11241]: So, we prefer a reduced form of QR while SVD for some reason is full :)

Comment (Lezcano) [reply aten/src/ATen/native/native_functions.yaml:11241]: I KNOW RIGHT??!!!! -.-

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 5]
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
...
-  const Tensor& R
-) {
-  return std::get<0>(linalg_qr_jvp(dA, Q, R));
-}
+    const auto sym = [](const Tensor& X) { return X + X.mH(); };
+    const auto syminv = [](const Tensor& X) {
+      auto ret = X.triu();
+      ret.diagonal(0, -2, -1).mul_(0.5);
+      return ret;
+    };
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:3206]: Fun to see that the input is const while the function has a side effect :)

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 6]
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
...
+  // Case m >= n
+  // We can put dQ in terms of dR
+  // dQ = dAR^{-1} - QdRR^{-1}
+  // Then we have
+  // Q^H dA R^{-1} = Q^HdQ + dRR^{-1}
+  // where Q^HdQ is skew Hermitian and dRR^{-1} is upper triangular
+  // Define sym(X) = X + X^H
+  // sym(dRR^{-1}) = sym(Q^H dA R^{-1})
+  // and define syminv(X) = triu(X) - 0.5 * diag(X) the inverse of sym : Triu(k) -> Her(k) to give
+  // dR = syminv(sym(Q^H dA R^{-1}))R
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:3172]: I think this works because the diagonal of `R` is real?

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3172]: Indeed. It read: "the inverse of `sym : Triu(k, diag \in R) -> \Her(k)`"

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 7]
```diff
@@ -3155,170 +3155,177 @@ std::tuple<Tensor, Tensor> linalg_lstsq_backward(
...
+  //
+  // Case m < n
+  // Put dR as a function of dQ
+  // dR = Q^H dA - Q^H dQ R
+  // Let X_1 be the main m x m submatrix of a matrix X \in C^{m x n}
+  // Q^H A_1 R_1^{-1} = Q^H dQ + dR_1 R_1^{-1}
+  // Define trilIm(X) = X.tril(-1) + i * Im diag(X)
+  // trilIm(Q^H dQ) = trilIm(Q^H A_1 R_1^{-1})
+  // and define trilIminv(X) = X - X^H - 0.5*i*Im diag(X). This is the inverse of
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:3181]: shouldn't `trillIminv(X) = X - X^H - i * im diag(X)`?

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3181]: Yup.

File: torch/linalg/__init__.py [hunk 8]
```diff
@@ -2320,20 +2321,17 @@
   When :attr:`mode`\ `= 'r'`, the `Q` tensor is an empty tensor.
   This behavior may change in a future PyTorch release.
 
-.. note:: The elements in the diagonal of `R` are not necessarily positive.
+.. note:: The only :attr:`mode` that is differentiable is :attr:`mode`\ `= 'reduced'`.
```
Comment (nikitaved) [inline torch/linalg/__init__.py:2325]: Shouldn't `mode=complete` with `m < n` be differentiable as well?

Comment (Lezcano) [reply torch/linalg/__init__.py:2325]: Yeah, the thing is that if `m < n`, mode complete and reduced are equivalent, so I think it's fine not advertising that really. You should be using `reduced`mode anyway.

**Comment by Lezcano** [comment]
@soulitzer I run `test_vmapvjpvjp` with the current changes (in-place `at::real(t.diagonal(0, -2, -1)).mul_(0.5);` and it passes. Leaving it like this for now, as it is more efficient.

File: torch/linalg/__init__.py (hunk 8 above)
Comment (nikitaved) [reply torch/linalg/__init__.py:2325]: cc @mruberry , @IvanYashchuk . What is your take on this one?

Comment (IvanYashchuk) [reply torch/linalg/__init__.py:2325]: I'm leaning towards being explicit about what's supported if we started writing this note, i.e. I would include that supported modes are "reduced" or "complete" && `m <= n`.

Comment (mruberry) [reply torch/linalg/__init__.py:2325]: Erring on the side of more detail sounds fine (and maybe an opportunity to suggest a best practice)

**Comment by Lezcano** [comment]
@nikitaved this is ready for another review.

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -10672,19 +10672,16 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
...
-           check_batched_gradgrad=False,
            supports_forward_ad=True,
            supports_fwgrad_bwgrad=True,
+           # In-place ops
+           check_batched_gradgrad=False,
+           sample_inputs_func=sample_inputs_linalg_qr_geqrf,
            skips=(
-               DecorateInfo(unittest.expectedFailure, 'TestCompositeCompliance', 'test_backward'),
+               # The test is wrong https://github.com/pytorch/pytorch/pull/76115/files#r854455360
                DecorateInfo(unittest.expectedFailure, 'TestCompositeCompliance', 'test_forward_ad'),
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:10682]: Could you please clarify? What is wrong exactly?

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:10682]: They don't handle strides properly in the test for empty tensors.

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:10682]: See the comment linked there for further discussion.

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:10682]: Sorry, I am missing something, but the link is to the PRs hash. Where do I find the comment?

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:10682]: Or do you mean the link that was removed in your PR regarding empty inputs?

**Review by nikitaved — APPROVED** [review]
Thank you, Mario. Looks great!

**Review by albanD — APPROVED** [review]
SGTM

**Comment by albanD** [comment]
The lint needs fixing before landing though!

**Pull request closed without merging**